### PR TITLE
8267239: C1: RangeCheckElimination for % operator if divisor is IntConstant

### PR DESCRIPTION
**Pull request by kelthuzadx**
% operator follows from this rule that the result of the remainder operation can be negative only if the dividend is negative, and can be positive only if the dividend is positive. Moreover, the magnitude of the result is always less than the magnitude of the divisor(See [LS 15.17.3](https://docs.oracle.com/javase/specs/jls/se8/html/jls-15.html#jls-15.17.3)).
      
So if `y` is a constant integer and not equal to 0, then we can deduce the bound of remainder operation:
-  x % -y  ==> [0, y - 1] RCE
-  x % y   ==> [0, y - 1] RCE
-  -x % y  ==> [-y + 1, 0]
-  -x % -y ==> [-y + 1, 0]

Based on above rationale, we can apply RCE for the remainder operations whose dividend is constant integer and >= 0,  e.g.:

```java
for(int i=0;i<1000;i++){
  int top5 = arr[i%5];  // Apply RCE if arr is a loop invariant
  ....
}
```

For more detailed RCE results, please check out the attachment on JBS, it was generated by ArithmeticRemRCE with additional flags -XX:+TraceRangeCheckElimination -XX:+PrintIR.

Testing:
- test/hotspot/jtreg/compiler/c1/(slowdebug)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8267239](https://bugs.openjdk.java.net/browse/JDK-8267239): C1: RangeCheckElimination for % operator if divisor is IntConstant

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Nils Eliasson](https://openjdk.java.net/census#neliasso) (@neliasso - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4083/head:pull/4083` \
`$ git checkout pull/4083`

Update a local copy of the PR: \
`$ git checkout pull/4083` \
`$ git pull https://git.openjdk.java.net/jdk pull/4083/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4083`

View PR using the GUI difftool: \
`$ git pr show -t 4083`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4083.diff">https://git.openjdk.java.net/jdk/pull/4083.diff</a>

</details>

### CODE REVIEW

**Comment by kelthuzadx** [comment]
/label add hotspot

The inference is not documented in JLS, so add hotspot-dev for more logical checking ;)

**Comment by neliasso** [comment]
Have you checked how C2 handles this case?

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
Looks like C2 does not implement this optimization (should be in `ModINode::Value`). We should add it.

File: src/hotspot/share/c1/c1_RangeCheckElimination.cpp [hunk 1]
```diff
@@ -228,7 +228,22 @@ void RangeCheckEliminator::Visitor::do_ArithmeticOp(ArithmeticOp *ao) {
...
+      // x % -y  ==> [0, y - 1] Apply RCE
+      // x % y   ==> [0, y - 1] Apply RCE
+      // -x % y  ==> [-y + 1, 0]
+      // -x % -y ==> [-y + 1, 0]
+      if (x_bound->has_lower() && x_bound->lower() >= 0) {
+        _bound = new Bound(0, NULL, y->type()->as_IntConstant()->value() - 1, NULL);
+      } else {
+        _bound = new Bound();
+      }
+    }else {
```
[the remark this answers is not in the document]
Comment (kelthuzadx) [reply src/hotspot/share/c1/c1_RangeCheckElimination.cpp]: > Looks like C2 does not implement this optimization (should be in ModINode::Value). We should add it.

Thanks for confirming the C2 part, I will double-check it and file an issue for that~

> Missing whitespace }else

Fixed.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me.

**Review by neliasso — APPROVED** [review]
Looks good.

**Comment by kelthuzadx** [comment]
Thank you Nils and Tobias for taking the time to review this patch! I've filed https://bugs.openjdk.java.net/browse/JDK-8267376 for investigating if it's possible to apply C2 RCE for % operator when divisor != 0.

/integrate

**Comment by kelthuzadx** [comment]
I think it is not easy/trivial to implement this for C2 since it idealize ModINode to other nodes if divisor is constant integer:

https://github.com/openjdk/jdk/blob/31139108c1ca9d355bd484d692830dfbc8317477/src/hotspot/share/opto/divnode.cpp#L946-L948

https://github.com/openjdk/jdk/blob/31139108c1ca9d355bd484d692830dfbc8317477/src/hotspot/share/opto/divnode.cpp#L166-L192

**Pull request closed without merging**